First pass at including adios2 config

pelahi opened the pull request:
Have updated the Adios interface and ImageAccessFactory so that an entry listing the adios2 configuration can be passed through the parset.

pelahi committed:
First pass at including adios2 config

Have updated the Adios interface and ImageAccessFactory so that an entry
listing the adios2 configuration can be passed through the parset.

alxndrwllmsn commented on askap/imageaccess/ADIOSImage.tcc:
```diff
@@ -95,16 +99,29 @@ void ADIOSImage<T>::makeNewTable(const casacore::TiledShape& shape, casacore::uI
   const casacore::uInt ndim = latShape.nelements();
 
   casacore::TableDesc description;
+  //PJE - why are we adding a hard-coded string???
```
This was a quick thing to make sure it works. Ideally, this will be passed via the config file or as an input parameter.

alxndrwllmsn commented on askap/imageaccess/ADIOSImage.tcc:
```diff
@@ -95,16 +99,29 @@ void ADIOSImage<T>::makeNewTable(const casacore::TiledShape& shape, casacore::uI
+  else 
+  {
+    // invoke configuration based call 
+    //PJE - question on whether we also need to include the operatorParams {{{"Varaible", "map"}}}
```
This will need to be in the config file or as an input to the storage manager. I think the variable needs an name...
In the long run, the operatorParams input is where we will put the compression parameters.

pelahi committed:
Merge branch 'adios_integration' into adios_integration_config

pelahi committed:
Adding configuration interface and parallel

- updating the ADIOS interface to make use of MPI_COMM_SELF when
  compiled with MPI but not run in collective fashion. This is
  preliminary.
- still fleshing out the best fashion to pass along configuration

alxndrwllmsn commented on askap/imageaccess/ADIOSImage.h:
```diff
@@ -98,6 +111,14 @@ template <class T> class ADIOSImage : public casacore::ImageInterface<T>
   casacore::LatticeRegion* regionPtr_p;
   casacore::uInt row_p;
   casacore::String config;
+  // PJE - not certain how best to store the adios configuration besides
```
There is an AdiosStMan constructor that takes a config filename (`Adios2StMan(MPI_Comm mpiComm, std::string configFile, from_config_t);`). Alternatively, we could pass the parset down and create our own subset of config parameters. This is more complicated but might be more consistent with the Askapsoft way of doing things.

alxndrwllmsn replied:
Nevermind, saw your changes below, disregard the first statement.

alxndrwllmsn commented on askap/imageaccess/ADIOSImage.h:
```diff
@@ -13,6 +13,12 @@
 //# Forward Declarations
 #include <casacore/casa/iosfwd.h>
 
+#ifdef ADIOS_HAS_MPI
```
Would this conditional have to be added to the CMake to switch this on?

alxndrwllmsn commented on askap/imageaccess/ADIOSImage.tcc:
```diff
@@ -107,20 +121,36 @@ void ADIOSImage<T>::makeNewTable(const casacore::TiledShape& shape, casacore::uI
   casacore::SetupNewTable newtab(filename, description, casacore::Table::New);
 
   // now invoke the Adios2 storage manager
+  // PJE - the issue here is that the constructors are only valid if ADIOS has not be compiled
```
This might be a naive question, but why is this?

alxndrwllmsn commented on askap/imageaccess/CasaADIOSImageAccessParallel.h:
```diff
@@ -0,0 +1,274 @@
+    /// @param[in] config, string pointing to adios configuration
+    template <class T> CasaADIOSImageAccessParallel<T>(askapparallel::AskapParallel &comms, std::string configname = "");
+
+    // PJE -- uncertain how many functions would really require overwriting so as to 
```
So far, the only ADIOSImage function that uses a comm is the makeNewTable function. This is only called via `CasaADIOSImageAccess<T>::create(...)' function.

This does raise the question about reading. The adios storage manager is used to read as well, but we don't need to initialise it for this as it's invoked automatically. Does this mean that it's using MPI for writing but not for reading? If so, do we need to somehow explicitly invoke it with a comm so we get parallel read as well?

pelahi commented:
Your questions regarding ADIOS and MPI are a good one. I could update the CMake to define this for compilation but I'd like something a little cleaner, like something present in the library that is defined. Still, for the moment, I'm going to update CMake by seeing if I can look at the adios config and see what it spits out.

pelahi committed:
Update for adios parallel

This is WIP for adding parallel adios access. This has not yet been
compiled or tested.

- Updates were to add ADIOS2_USE_MPI defs around the adios access. Not
  clear if more defs around code is needed
- Added the parallel accessors class for adios for collective
  operations.

alxndrwllmsn committed:
fixed some typos and added config passthrough

alxndrwllmsn committed:
improved setup of cube

alxndrwllmsn committed:
added test with config

alxndrwllmsn commented:
@pelahi I've added some fixes and tests and ensured the config side of things works. Have a look and we can merge it into adios_integration if you're happy with it. We can then create a new branch for the MPI shenanigans.

pelahi commented:
I happy with this. Let's merge.
